### DISCUSSION TITLE
Where best to query DB during sign-in flow

### DISCUSSION THREAD

**Discussion by Toryn02** [nextauthjs/next-auth#3505] [Help] [answered]
Hi everyone, 

I'm new to NextAuth and I've been reading through the docs, the source code, and GitHub discussion forums, but I can't seem to find an answer to my question below.  I'm using NextAuth's Google sign-in provider and here's my desired use case:

1) User clicks on "Sign in with Google" button to begin sign-in flow.
2) NextAuth's Google provider handles user authentication and returns the relevant user data.
3) **Before the user is redirected (i.e., before any sign-in success message/UI change),** query both Firestore and Stripe to see if an existing user/customer record matches the email of the user who just signed in and, if not, create a new Firestore user record and Stripe customer record.
4) Add both the user's Firestore user record ID and Stripe customer ID to their client-side session object.
5) Redirect the user to the URL where they initiated the sign-in process and update the UI accordingly.

I've got steps 1, 2, 4, and 5 above working fine, but I'd very much love any help that anyone can provide with where best in the NextAuth sign-in flow to accomplish step 3 above (e.g., is it in the "jwt" callback?).  If I'm missing something in the documentation, please let me know.

Thank you so much, and I really appreciate it!

Regards,
Josh

**Comment by balazsorban44** [comment, marked as answer]
I believe this is sufficiently documented, but correct me if I'm wrong.

1,2 and 5 are out of the box, you don't need to do anything other than calling `signIn("google")` on the page you want the user to be logged in at. (you can configure where to return by the `callbackUrl` option in the second argument, please check the docs).

For 3, I would recommend the `signIn` callback.
For 4, I would recommend the `jwt` callback.


https://next-auth.js.org/getting-started/client#specifying-a-callbackurl
https://next-auth.js.org/configuration/callbacks

**Reply by Toryn02** [reply]
Thank you, @balazsorban44, for taking the time to reply!  After some further experimentation, the `signIn()` callback is indeed the best place to handle my step #3.  

The thing that confused me in the documentation is that the `signIn()` callback is introduced with the single sentence: "Use the `signIn()` callback to control if a user is allowed to sign in."  Since this wasn't my use case, I kept looking elsewhere in the docs for the solution.

I'd suggest perhaps expanding that sentence to read something along the lines of: "The `signIn()` callback can be used to handle a variety of post-authentication actions prior to redirect, such as creating or updating a user's database record or controlling whether a user is allowed to sign in."  Just my two cents.  Thanks again!

**Reply by balazsorban44** [reply]
The intention is really to only check if the user is allowed to log in or not, but currently that's the best place for what you are trying to do as well. We might iterate on this later on.

There is also events (where you have a signin event), but I'm not sure in which state that is called from the top of my head: https://next-auth.js.org/configuration/events#signin

**Comment by Toryn02** [comment]
Understood, and thanks again!
